referenceframe/urdf.go

```diff
@@ -0,0 +1,313 @@
+
+	// Migrate URDF elements into an equivalent ModelConfig representation
+	mc.Name = modelName
+	mc.KinParamType = "SVA"
+
+	// Handle joints
+	for _, jointElem := range urdf.Joints {
+		// Checking for reserved names in this or adjacent elements
+		if jointElem.Name == World {
+			return nil, errors.New("Joints with the name 'world' are not supported by config parsers")
+		}
+
+		// Relationship tracking
+		parentMap[jointElem.Name] = jointElem.Parent.Link
+		parentMap[jointElem.Child.Link] = jointElem.Name
+
+		// Set up the child link mentioned in this joint; fill out the details in the link parsing section later
+		childLink := JsonLink{ID: jointElem.Child.Link, Parent: jointElem.Name}
+
+		switch jointElem.Type {
+		case "continuous", "revolute", "prismatic":
+			// Parse important details about each joint, including axes and limits
+			jointAxes := convStringAttrToFloats(jointElem.Axis.XYZ)
+			thisJoint := JsonJoint{
+				ID:     jointElem.Name,
+				Type:   jointElem.Type,
+				Parent: jointElem.Parent.Link,
+				Axis:   spatial.AxisConfig{jointAxes[0], jointAxes[1], jointAxes[2]},
+			}
+
+			// Slightly different limits handling for continuous, revolute, and prismatic joints
+			if jointElem.Type == "continuous" {
+				thisJoint.Type = "revolute" // Currently, we treate a continuous joint as a special case of the revolute joint
+				thisJoint.Min, thisJoint.Max = math.Inf(-1), math.Inf(1)
+			} else if jointElem.Type == "prismatic" {
+				thisJoint.Min, thisJoint.Max = metersToMM(jointElem.Limit.Lower), metersToMM(jointElem.Limit.Upper)
+			} else if jointElem.Type == "revolute" {
+				thisJoint.Min, thisJoint.Max = utils.RadToDeg(jointElem.Limit.Lower), utils.RadToDeg(jointElem.Limit.Upper)
+			} else {
+				return nil, err
+			}
+
+			mc.Joints = append(mc.Joints, thisJoint)
+
+			// Generate child link translation and orientation data, which is held by this joint per the URDF design
+			childXYZ := convStringAttrToFloats(jointElem.Origin.XYZ)
+			childRPY := convStringAttrToFloats(jointElem.Origin.RPY)
+			childEA := spatial.EulerAngles{Roll: childRPY[0], Pitch: childRPY[1], Yaw: childRPY[2]}
+			childOrient, err := spatial.NewOrientationConfig(childEA.EulerAngles())
+
+			// Note the conversion from meters to mm
+			childLink.Translation = spatial.TranslationConfig{
+				metersToMM(childXYZ[0]),
+				metersToMM(childXYZ[1]),
+				metersToMM(childXYZ[2]),
+			}
+			childLink.Orientation = *childOrient
+
+			if err != nil {
+				return nil, err
+			}
+		case "fixed":
+			// Handle fixed joint -> static link conversion instead of adding to Joints[]
+			thisLink := JsonLink{ID: jointElem.Name, Parent: jointElem.Parent.Link}
+
+			linkXYZ := convStringAttrToFloats(jointElem.Origin.XYZ)
+			linkRPY := convStringAttrToFloats(jointElem.Origin.RPY)
+			linkEA := spatial.EulerAngles{Roll: linkRPY[0], Pitch: linkRPY[1], Yaw: linkRPY[2]}
+			linkOrient, err := spatial.NewOrientationConfig(linkEA.AxisAngles())
```

Comment:
same here.  Actually it seems like there is a lot of duplicate code here and the above case that could be abstracted away